google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreOptions.java

```diff
@@ -82,53 +88,123 @@ protected boolean projectIdRequired() {
-    return DEFAULT_HOST;
+    return FirestoreDefaults.INSTANCE.getHost();
   }
 
   public String getDatabaseId() {
     return databaseId;
   }
 
+  public CredentialsProvider getCredentialsProvider() {
+    return credentialsProvider;
+  }
+
+  public TransportChannelProvider getTransportChannelProvider() {
+    return channelProvider;
+  }
+
   public static class Builder extends ServiceOptions.Builder<Firestore, FirestoreOptions, Builder> {
 
-    private String databaseId = DEFAULT_DATABASE_ID;
+    private String databaseId = FirestoreDefaults.INSTANCE.getDatabaseId();
+    private TransportChannelProvider channelProvider =
+        FirestoreDefaults.INSTANCE.getTransportChannelProvider();
+    private CredentialsProvider credentialsProvider =
+        FirestoreDefaults.INSTANCE.getCredentialsProvider();
 
     private Builder() {}
 
     private Builder(FirestoreOptions options) {
       super(options);
+      this.databaseId = options.databaseId;
+      this.channelProvider = options.channelProvider;
+      this.credentialsProvider = options.credentialsProvider;
     }
 
+    /**
+     * Sets the {@link TransportOptions} to use with this Firestore client.
+     *
+     * @param transportOptions A GrpcTransportOptions object that defines the transport options for
+     * this client.
+     */
     @Nonnull
     @Override
     public Builder setTransportOptions(@Nonnull TransportOptions transportOptions) {
       if (!(transportOptions instanceof GrpcTransportOptions)) {
         throw new IllegalArgumentException(
-            "Only grpc transport is allowed for " + API_SHORT_NAME + ".");
+            "Only GRPC transport is allowed for " + API_SHORT_NAME + ".");
       }
       return super.setTransportOptions(transportOptions);
     }
 
-    @Override
+    /**
+     * Sets the {@link TransportChannelProvider} to use with this Firestore client.
+     *
+     * @param channelProvider A InstantiatingGrpcChannelProvider object that defines the transport
+     * provider for this client.
+     */
     @Nonnull
-    public FirestoreOptions build() {
-      return new FirestoreOptions(this);
+    public Builder setChannelProvider(@Nonnull TransportChannelProvider channelProvider) {
+      if (!(channelProvider instanceof InstantiatingGrpcChannelProvider)) {
+        throw new IllegalArgumentException(
+            "Only GRPC channels are allowed for " + API_SHORT_NAME + ".");
+      }
+      this.channelProvider = channelProvider;
+      return this;
     }
 
+    /**
+     * Sets the {@link CredentialsProvider} to use with this Firestore client.
+     *
+     * @param credentialsProvider A GoogleCredentialsProvider object that defines the credential
+     * provider for this client.
+     */
+    @Nonnull
+    public Builder setCredentialsProvider(@Nonnull CredentialsProvider credentialsProvider) {
+      if (!(credentialsProvider instanceof GoogleCredentialsProvider)) {
+        throw new IllegalArgumentException(
```

Comment:
Out of curiosity, why do we have this restriction?

Reply:
I looked through the callers and it seems like it is not needed.